autox/informer.go

```diff
@@ -3,154 +3,365 @@ package autox
+			if err1 := installHelmRelease(releaseName, group, releaseSpec, namespace); err1 != nil {
+				err = errors.New("one or more Helm release installs failed")
+			}
+		case deleteAction:
+			// if there is an error, keep going forward in the for loop
+			if err1 := deleteHelmRelease(releaseName, group, namespace); err1 != nil {
+				err = errors.New("one or more Helm release deletions failed")
 			}
 		}
-	} else {
-		// TODO: what log level should this be?
-		log.Logger.Debug("AutoX should make a Helm release for app \"", resourceName, "\" but no Helm charts were provided in the chartGroupConfig")
 	}
-}
-
-var addObject = func(obj interface{}) {
-	log.Logger.Debug("Add:", obj)
-
-	uObj := obj.(*unstructured.Unstructured)
-	resourceName := uObj.GetName()
-	// example label: iter8.tools/autox-group=hello
-	labels := pruneLabels(uObj.GetLabels())
 
-	// check if the app name matches the name in the autox label
-	if autoxLabelName, ok := labels[autoxLabel]; ok && resourceName == autoxLabelName {
-		// Release Helm charts
-		installHelmReleases(resourceName)
+	if err != nil {
+		log.Logger.Error(err)
 	}
+
+	return err
 }
 
+// pruneLabels will extract the labels that are relevant for autoX
 func pruneLabels(labels map[string]string) map[string]string {
-	// TODO: select labels important for autoX
-
-	return labels
+	prunedLabels := map[string]string{}
+	for _, l := range []string{autoXLabel, appLabel, versionLabel, trackLabel} {
+		prunedLabels[l] = labels[l]
+	}
+	return prunedLabels
 }
 
-var updateObject = func(oldObj, obj interface{}) {
-	log.Logger.Debug("Update:", oldObj, obj)
+// hasAutoXLabel checks if autoX label is present
+func hasAutoXLabel(labels map[string]string) bool {
+	_, ok := labels[autoXLabel]
+	return ok
+}
 
-	uOldObj := oldObj.(*unstructured.Unstructured)
-	oldLabels := pruneLabels(uOldObj.GetLabels())
+// handle is the entry point to all (add, update, delete) event handlers
+func handle(obj interface{}, releaseGroupSpecName string, releaseGroupSpec releaseGroupSpec) {
+	m.Lock()
+	defer m.Unlock()
 
-	uObj := obj.(*unstructured.Unstructured)
-	resourceName := uObj.GetName()
-	// example label: iter8.tools/autox-group=hello
-	labels := pruneLabels(uObj.GetLabels())
+	// parse object
+	u := obj.(*unstructured.Unstructured)
 
-	// if the pruned labels are the same, do nothing
-	if reflect.DeepEqual(oldLabels, labels) {
+	// check if name matches trigger
+	name := u.GetName()
+	if name != releaseGroupSpec.Trigger.Name {
 		return
 	}
 
-	// if the pruned labels are different, then update by deleting and reinstalling
-	deleteHelmReleases(resourceName)
-	installHelmReleases(resourceName)
+	// namespace and GVR should already match trigger
+	ns := u.GetNamespace()
+	// Note: GVR is from the release group spec, not available through the obj
+	gvr := getGVR(releaseGroupSpec)
+
+	// always delete Helm releases
+	log.Logger.Debugf("delete Helm releases for release group \"%s\"", releaseGroupSpecName)
```

Comment:
There are situations when deleting always is not desirable. For e.g., consider if your trigger object is a deployment, and its status has changed. We shouldn't recreate the experiment in this instance.

Perhaps delete if the object no longer exists in the cluster, or if the required autox labels have been removed? If not, simply doChartAction(..., create/update, ...) instead?